hibernate-core/src/main/java/org/hibernate/query/sqm/internal/AbstractSqmSelectionQuery.java

```diff
@@ -166,8 +166,13 @@ private SqmSelectStatement<KeyedResult<R>> paginateQuery(
 						getSqmSelectStatement().copy( noParamCopyContext() );
 		final NodeBuilder builder = sqm.nodeBuilder();
 		//TODO: find a better way handle parameters
-		builder.setCriteriaValueHandlingMode(ValueHandlingMode.INLINE);
-		return paginate( keyDefinition, keyValues, sqm, builder );
+		final ValueHandlingMode valueHandlingMode = builder.setCriteriaValueHandlingMode(ValueHandlingMode.INLINE);
```

Comment:
Is there a reason why you need to do this? Changing the value handling mode affects the whole JVM, so this is a concern in a multi-threaded environment.

Reply:
Yes, it's issue https://hibernate.atlassian.net/browse/HHH-18596, due to a problem with the query interpretation cache that I mentioned to you a long time ago. I would _love_ to see that fixed so that we can get rid of this and close that issue.

Reply:
Well, I guess I will look into this next then, since this is a pretty severe issue IMO.

Reply:
Awesome, that would be great.